xdsl/dialects/riscv.py

```diff
@@ -223,6 +223,41 @@ def __init__(
         )
 
 
+class RdImmJumpOperation(IRDLOperation, RISCVOp, ABC):
+    """
+    In the RISC-V spec, this is the same as `RdImmOperation`. For jumps, the `rd` register
+    is neither an operand, because the stored value is overwritten, nor a result value,
+    because the value in `rd` is not defined after the jump back. So the `rd` makes the
+    most sense as an attribute.
```

Comment:
Hi @superlopuh, I am not sure why you feel `rd` is not a real result value and worth introducing an entirely new kind of register concept. AFAIU the rd register will be consumed/used by a return instruction, but the jump instructions write it like any normal register. Or do I miss sth?

Also, would you not at the very least need to update the register allocator to support registers stored as attributes, but no the types.

Reply:
It seems like an attribute is the most appropriate place because it doesn't quite fit as an operand or result value. It doesn't fit as an operand, because the value in the register is overwritten by `pc+4`, and it doesn't fit as a result, as the value is not defined when control comes back to the next instruction.

For our current register allocation approach, it should work fine, since a `None` value will just not be printed in the assembly, meaning the default `ra` will be used by the emulator. When we replace our allocator with something more sophisticated, it would have to take this into account, but I would assume that it would have to be much more sophisticated to mess with custom return addresses.

Reply:
I think the distinction here is that the SSAValue that is stored in the register passed as `rd` is not consumed. In fact, it may not be defined before the call, so it's not quite an operand. Similarly, the content of the register when control returns, is not well defined, so to use that register as an SSAValue is not defined either.

Reply:
I feel the rd could be represented as a result too, holding the value of the register after a "return" if any happens. It might not make sense in what we want to do with the dialect now, but I don't see a reason not to allow it in the dialect in general.
We can make it optional (as Sasha did after discussion) to signify when it's just ignored. I don't know what the actual register allocation implementation will look like, but I feel that making operands and results optional in general might make its work easier?

Reply:
In our discussion I forgot about the register allocation, and it's made more complex by having it be an optional result. I think the attribute is the way to go for the time being. To be honest this is also a really weird corner case, and I'm not sure that it's worth spending a lot of time on it now.